vm/src/vm/decoding/decoder.rs

```diff
@@ -102,24 +102,26 @@ pub fn decode_instruction(encoded_instr: u128) -> Result<Instruction, VirtualMac
 
     let opcode_extension = match opcode_extension_num {
         0 => OpcodeExtension::Stone,
-        1 => {
-            if opcode != Opcode::NOp
-                || (op1_addr != Op1Addr::FP && op1_addr != Op1Addr::AP)
-                || res != Res::Op1
-                || pc_update != PcUpdate::Regular
-                || (ap_update_num != 0 && ap_update_num != 2)
-            {
-                return Err(VirtualMachineError::InvalidBlake2sFlags(flags & 0x7FFF));
-            };
-            OpcodeExtension::Blake
-        }
+        1 => OpcodeExtension::Blake,
+        2 => OpcodeExtension::BlakeFinalize,
         _ => {
             return Err(VirtualMachineError::InvalidOpcodeExtension(
                 opcode_extension_num,
             ))
         }
     };
 
+    if (opcode_extension == OpcodeExtension::Blake
+        || opcode_extension == OpcodeExtension::BlakeFinalize)
+        && (opcode != Opcode::NOp
+            || (op1_addr != Op1Addr::FP && op1_addr != Op1Addr::AP)
+            || res != Res::Op1
+            || pc_update != PcUpdate::Regular
+            || (ap_update_num != 0 && ap_update_num != 2))
+    {
```

Comment:
I would extract these conditionals into a variable to separate the `opcode_extension` checks of the flags check. Something like:

```suggestion
    let are_blake_flags_invalid = opcode != Opcode::NOp
            || (op1_addr != Op1Addr::FP && op1_addr != Op1Addr::AP)
            || res != Res::Op1
            || pc_update != PcUpdate::Regular
            || (ap_update_num != 0 && ap_update_num != 2;

    if (opcode_extension == OpcodeExtension::Blake
        || opcode_extension == OpcodeExtension::BlakeFinalize)
        && are_opcodes_flags_invalid)
    {
```